autonav_ws/src/autonav_hardware/src/can_node.py

```diff
@@ -0,0 +1,289 @@
+        super.__init__("CAN_node")
+        self.write_config(CanConfig)
+
+        # can
+        self.can = None
+
+        # safety lights
+        self.safetyLightsSubscriber = self.create_subscription(
+            SafetyLights,
+            "autonav/safety_lights",
+            self.on_safety_lights_received(),
+            20
+        )
+
+        # motor messages
+        self.motorInputSubscriber = self.create_subscription(
+            MotorInput,
+            "autonav/motor_input",
+            self.on_motor_input_received(),
+            20
+        )
+
+        self.motorFeedbackPublisher = self.create_publisher(
+            MotorFeedback,
+            "/autonav/motor_feedback", 
+            20
+        )
+
+        # conbus
+        self.conbusSubscriber = self.create_subscription(
+            Conbus,
+            "/autonav/conbus/instruction", 
+            self.on_conbus_received(), 
+            20
+        )
+
+        self.conbusPublisher = self.create_publisher(
+            Conbus, 
+            "/autonav/conbus/data", 
+            20
+        )
+
+        # motor statistics and PID tuning
+        self.linearPIDStatisticsPublisher = self.create_publisher(
+            LinearPIDStatistics,
+            "autonav/linear_pid_statistics",
+            20
+        )
+        self.angularPIDStatisticsPublisher = self.create_publisher(
+            AngularPIDStatistics,
+            "autonav/angular_pid_statistics",
+            20
+        )
+        self.motorStatisticsFrontMotorsPublisher = self.create_publisher(
+            MotorStatisticsFrontMotors,
+            "autonav/motor_statistics_front_motors",
+            20
+        )
+        self.motorStatisticsBackMotorsPublisher = self.create_publisher(
+            MotorStatisticsBackMotors,
+            "autonav/motor_statistics_back_motors",
+            20
+        )
+
+
+    def init(self):
+        # can threading
+        self.canTimer = self.create_timer(0.5, self.can_worker)
+        self.canReadThread = threading.Thread(target=self.can_thread_worker)
+        self.canReadThread.daemon = True
+
+
+    def can_worker(self):
+        try:
+            with open("/dev/ttyACM0", "r") as f:
+                pass
+
+            if self.can is not None:
+                return
+
+            self.can = can.ThreadSafeBus(
+                bustype="slcan", channel="/dev/ttyACM0", bitrate=100000)
+            self.set_device_state(DeviceState.OPERATING)
+        except:
+            if self.can is not None:
+                self.can = None
+
+            if self.device_state != DeviceState.WARMING:
+                self.set_device_state(DeviceState.WARMING)
+
+
+    def can_thread_worker(self):
+        while rclpy.ok():
+            if self.device_state != DeviceState.READY and self.device_state != DeviceState.OPERATING:
+                continue
+            if self.can is not None:
+                try:
+                    msg = self.can.recv(timeout=0.01)
```

Comment:
is this the same timeout as last year? are we sure it's good? just wondering if this might be the root cause of last year's CAN weirdness, or if it'd be better to make it longer or shorter or something